modules/core/arrow-core/src/main/kotlin/arrow/core/TupleN.kt

```diff
@@ -135,4 +136,96 @@ data class Tuple22<out A, out B, out C, out D, out E, out F, out G, out H, out I
   companion object
 }
 
+private const val INT_MAX_POWER_OF_TWO: Int = Int.MAX_VALUE / 2 + 1
+
 infix fun <A, B> A.toT(b: B): Tuple2<A, B> = Tuple2(this, b)
+
+fun <K, V> Tuple2<K, V>.toPair() : Pair<K, V> = Pair(this.a, this.b)
+
+fun <K, V> Pair<K, V>.toTuple2() : Tuple2<K, V> = Tuple2(this.first, this.second)
+
+fun <K, V> mapOf(vararg tuples: Tuple2<K, V>): Map<K, V> =
+  if (tuples.isNotEmpty()) tuples.toMap(LinkedHashMap(mapCapacity(tuples.size))) else emptyMap()
+
+fun <K, V> Iterable<Tuple2<K, V>>.toMap(): Map<K, V> {
+  if (this is Collection) {
+    return when (size) {
+      0 -> emptyMap()
+      1 -> mapOf(if (this is List) this[0] else iterator().next())
+      else -> toMap(LinkedHashMap(mapCapacity(size)))
+    }
+  }
+  return toMap(LinkedHashMap()).optimizeReadOnlyMap()
+}
+
+fun <K, V> Array<out Tuple2<K, V>>.toMap(): Map<K, V> = when (size) {
+  0 -> emptyMap()
+  1 -> mapOf(this[0])
+  else -> toMap(LinkedHashMap<K, V>(mapCapacity(size)))
+}
+
+fun <K, V> Sequence<Tuple2<K, V>>.toMap(): Map<K, V> = toMap(LinkedHashMap()).optimizeReadOnlyMap()
+
+fun <K, V> mapOf(pair: Tuple2<K, V>): Map<K, V> = Collections.singletonMap(pair.a, pair.b)
+
+internal fun <K, V, M : MutableMap<in K, in V>> Iterable<Tuple2<K, V>>.toMap(destination: M): M =
```

Comment:
what are these `internal` ones needed for?

Reply:
They are handling the MutableMaps. The concrete implementation being used in this PR is LinkedHashMap (this is the same as Kotlin standard library), so it need a function to deal with this MutableMap. 

Basically is creates the MutableMap and adds all the object in the Iterable/Array/Sequence in it, but the PR only exposes the Map (unmutable).